### ISSUE TITLE
[android] Direct device-to-device sharing of map files

### ISSUE THREAD

**Issue by rtsisyk** [organicmaps/organicmaps#1801]
# Problem

I am hearing countless complaints that the maps directory is not accessible anymore. Unfortunately, we have some technical limitation in Android 11+. There is no convenient way to use the shared storaged anymore. 

# Solution

The typical use-case is to transfer maps files from device to another without having the Internet connection. For example, you have a map downloaded, but your fellow traveler doesn't. Instead of thinking about raw access to the private storage, we can implement direct device-to-device sharing.

The workflow is the following:

1. Open the list of maps on your device
2. Select one particular map or entire country
3. Click "Share" button
4. Android will show the system sharing dialog with multiple options
5. On the second device, receive a file and open it via Organic Maps app
6. The app will open the map file and import it into the private storage

The beatify of this approach is that users will have numerous options to exchange maps files:

- Bluetooth
- Nearby Share
- Quick Share
- Save as file
- WhatsApp
- Telegram

Plus it provides a convinient method to test custom/new maps.

CC: @biodranik @vng

**Comment by biodranik** [comment]
I don't think that you named the main problem. From our users and from the past, the more actual problem was this one:

- There is no Internet access on the device (or it is very slow).
- APK is available and installed.
- Maps are downloaded manually on some PC using the mirror (or are brought on a USB stick).

The goal is to copy (import) pre-downloaded maps on the device somehow. Do you see any solutions for that?

**Comment by rtsisyk** [comment]
> The goal is to copy (import) pre-downloaded maps on the device somehow. Do you see any solutions for that?

Sorry for the late answer. Sure, my proposal solves exactly this problem. Android supports numerous ways to share files via Sharing menu. Some of them don't require the Internet connection at all (e.g. Bluetooth and Nearby Sharing). You can transfer map files from one device to another easily without need to have the storage access.

**Comment by batyrmastyr** [comment]
Android to android sharing is nice to have but Android to iOS sharing is also needed. Use case: I have iPhone, my wife prefers Android and we usually forget to download

**Comment by biodranik** [comment]
That in theory could be done via a Bluetooth Sync.

**Comment by pastk** [comment]
Technically, another way to share map files will be by implementing a [DocumentsProvider](https://developer.android.com/guide/topics/providers/create-document-provider) for them - hence the files will be visible and copiable via a file manager. Its just an assumption based on reading the docs so far :)

**Comment by endim8** [comment]
For anyone trying to load mwms with root access, the maps are stored at /storage/emulated/0/Android/data/app.organicmaps/files/{6 numbers - map generation date}

**Comment by 0709wiwiwi** [comment]
> For anyone trying to load mwms with root access, the maps are stored at

Given Android is increasingly closing down access to data, apps like Cruiser & Locus offer alternative storage choices into media (internal or extsd). The access control drift in here still appears to be slightly more tolerant so far. 
Example: /storage/1BA0-1D11/Android/media/gr.talent.cruiser/maps

**Comment by biodranik** [comment]
These apps have unsafe android.permission.WRITE_EXTERNAL_STORAGE permission. If you use external storage in OM settings, in most cases you usually are able to access maps using adb.

**Comment by 0709wiwiwi** [comment]
Someone with too much fear of the outside world should already better not own any electronic device ?

**Comment by patepelo** [comment]
> I have installed Organic Maps on two devices.
> 
> To save on data downloads, I would like to share the map data that I have downloaded or updated on my phone with my tablet. Both devices are now running Android 14.
> 
> I would therefore like an export and import function for map data, similar to what is available for tracks and favorites.
> 
> Note: OSMAnd has already implemented this feature.
> 
> With Android’s built-in file manager, it is possible to copy the map data to another device, but due to Android restrictions, it cannot be written back into the correct directory on the second device.

#Emails&Reviews
Subject: [2025.01.28-1-Google] Organic Maps Bugreport
Timestamp: Tue, 04 Feb 2025 1:35:13 PM -0400
